app/bundles/EmailBundle/Entity/EmailRepository.php

```diff
@@ -574,6 +576,83 @@ public function upCount($id, $type = 'sent', $increaseBy = 1, $variant = false)
         }
     }
 
+    public function upCountSent(int $id, int $increaseBy = 1, bool $variant = false): void
+    {
+        if ($increaseBy <= 0) {
+            return;
+        }
+
+        $connection  = $this->getEntityManager()->getConnection();
+        $updateQuery = $connection->createQueryBuilder()
+            ->update(MAUTIC_TABLE_PREFIX.'emails')
+            ->set('sent_count', 'sent_count + :increaseBy')
+            ->where('id = :id');
+
+        if ($variant) {
+            $updateQuery->set('variant_sent_count', 'variant_sent_count + :increaseBy');
+        }
+
+        $updateQuery
+            ->setParameter('increaseBy', $increaseBy)
+            ->setParameter('id', $id);
+
+        // Try to execute 3 times before throwing the exception
+        $retrialLimit = 3;
+        while ($retrialLimit >= 0) {
+            try {
+                $updateQuery->executeStatement();
+
+                return;
+            } catch (\Doctrine\DBAL\Exception $e) {
+                --$retrialLimit;
+                if (0 === $retrialLimit) {
+                    throw $e;
+                }
+            }
+        }
+    }
+
+    public function incrementRead(int $emailId, int $statId, bool $isVariant = false): void
+    {
+        $q = $this->getEntityManager()->getConnection()->createQueryBuilder();
+
+        $subQuery = $this->getEntityManager()->getConnection()->createQueryBuilder()
+            ->select('es.email_id')
+            ->from(MAUTIC_TABLE_PREFIX.'email_stats', 'es')
+            ->where('es.id = :statId')
+            ->andWhere('es.is_read = 1');
+
+        $q->update(MAUTIC_TABLE_PREFIX.'emails', 'e')
+            ->set('read_count', 'read_count + 1')
+            ->where(
+                $q->expr()->and(
+                    $q->expr()->eq('e.id', ':emailId'),
+                    $q->expr()->notIn('e.id', $subQuery->getSQL())
```

Comment:
Can you give me a hit why we have to exclude emails that have `es.is_read = 1`? What's so special about `1`?

Reply:
This means that the concurrent request has already incremented the email read counter. So at this point we want to make sure that the es is not marked as read.

Reply:
But this will work only for the first read, right? How about the second when the is_read flag is set? The read_count will never be increased above 1.

Reply:
But this method is not called on email send but on email hit. I don't understand how can this work on the second time the email is read?

Reply:
Perhaps it would be better to replace the unit tests with a functional test that will open the email twice to see if this works. 

I've written some unit tests asserting raw SQL queries in the past and found that it's a pain to maintain and it does not tell you if the feature works or not. So they add more burden to the project and add no value.

Reply:
Yes, for example: we send 1 email to 3 contacts, so 3 `email_stats` are created. Each contact can only increment the email read counter once because they call this method with parameters:
1. incrementRead(1, 1)
2. incrementRead(1, 2)
3. incrementRead(1, 3)

The SQL query looks like this:
```
UPDATE test_emails e SET read_count = read_count + 1 WHERE (e.id = :emailId) AND (e.id NOT IN (SELECT es.email_id FROM test_email_stats es WHERE (es.id = :statId) AND (es.is_read = 1)))
```

This works, the read_count will be increased above 1, but only once per contact (email sent to be precise).

Reply:
We can make a test to ensure that the stat will be increased above 1, but i don't think we can write a test to cover the concurrency issue because it will be unstable.

Reply:
Now I understand. Thanks for the example! So you are saying we want only unique email read count in the `emails` table. It used to be overall count before (+- concurrency inaccuracy)

Reply:
It is already unique, but it's verified before the update in the EmailModel (https://github.com/mautic/mautic/blob/bfbf0a5dc58d12ba346c9ba58b89413b48db6771/app/bundles/EmailBundle/Model/EmailModel.php#L515). I've tested this and I was able only increment this counter once per email sent (without using concurrency). So it's not BC, only additional check while doing the update.

Reply:
Thanks for clearing my doubts! The code makes sense.